src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs

```diff
@@ -1194,23 +1181,38 @@ public void LoadKeysLookupLock()
         {
             lock (this.lockObject)
             {
-                var lookup = new Dictionary<Script, HdAddress>();
-                foreach (var wallet in this.Wallets)
+                //var lookup = new Dictionary<Script, HdAddress>();
+                foreach (Wallet wallet in this.Wallets)
                 {
-                    var accounts = wallet.GetAccountsByCoinType(this.coinType);
-                    foreach (var account in accounts)
+                    IEnumerable<HdAddress> addresses = wallet.GetAllAddressesByCoinType(this.coinType);
+                    foreach (HdAddress address in addresses)
                     {
-                        var addresses = account.ExternalAddresses.Concat(account.InternalAddresses);
-                        foreach (var address in addresses)
-                        {
-                            lookup.Add(address.ScriptPubKey, address);
-                            if (address.Pubkey != null)
-                                lookup.Add(address.Pubkey, address);
-                        }
+                        this.keysLookup[address.ScriptPubKey] = address;
+                        if (address.Pubkey != null)
+                            this.keysLookup[address.Pubkey] = address;
                     }
                 }
+            }
+        }
+
+        /// <summary>
+        /// Update the keys and transactions we're tracking in memory for faster lookups.
+        /// </summary>
+        public void UpdateKeysLookupLock(IEnumerable<HdAddress> addresses)
```

Comment:
I would suggest (for the sake of avoiding double locks) to make this methods private `UpdateKeysLookupLock` `LoadKeysLookupLock` (if needed for unit testing then a public Internal method) and remove the lock from inside then make sure the callers are holding the lock, there are only 6 places where this method is called.

Reply:
Leave this to the wallet refactor when we make it stand alone I think this is good as it is now.